tutorials/publish_subscribe/c/home_automation_subscriber_with_timestamp.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * (c) 2024 Copyright, Real-Time Innovations, Inc.  All rights reserved.
+ *
+ *  RTI grants Licensee a license to use, modify, compile, and create derivative
+ *  works of the Software.  Licensee has the right to distribute object form
+ *  only for use with RTI products.  The Software is provided "as is", with no
+ *  warranty of any type, including any warranty for fitness for any purpose.
+ *  RTI is under no obligation to maintain or support the Software.  RTI shall
+ *  not be liable for any incidental or consequential damages arising out of the
+ *  use or inability to use the software.
+ */
+
+#include "ndds/ndds_c.h"
+#include "home_automation.h"
+#include "home_automationSupport.h"
+
+static int subscriber_shutdown(
+    DDS_DomainParticipant *participant)
+{
+    DDS_ReturnCode_t retcode;
+    int status = 0;
+
+    if (participant != NULL) {
+        retcode = DDS_DomainParticipant_delete_contained_entities(participant);
+        if (retcode != DDS_RETCODE_OK) {
+            fprintf(stderr, "delete_contained_entities error %d\n", retcode);
+            status = -1;
+        }
+
+        retcode = DDS_DomainParticipantFactory_delete_participant(
+            DDS_TheParticipantFactory, participant);
+        if (retcode != DDS_RETCODE_OK) {
+            fprintf(stderr, "delete_participant error %d\n", retcode);
+            status = -1;
+        }
+    }
+
+    return status;
+}
+
+void DeviceStatusListener_on_data_available(
+    void* listener_data,
+    DDS_DataReader* reader)
+{
+    DeviceStatusDataReader *DeviceStatus_reader = NULL;
+    struct DeviceStatusSeq data_seq = DDS_SEQUENCE_INITIALIZER;
+    struct DDS_SampleInfoSeq info_seq = DDS_SEQUENCE_INITIALIZER;
+    struct DDS_Time_t timestamp;
+    DeviceStatus *device_status = NULL;
+    DDS_ReturnCode_t retcode;
+    int i;
+
+
+    DeviceStatus_reader = DeviceStatusDataReader_narrow(reader);
+    if (DeviceStatus_reader == NULL) {
+        fprintf(stderr, "DataReader narrow error\n");
+        return;
+    }
+
+    retcode = DeviceStatusDataReader_take(
+            DeviceStatus_reader,
+            &data_seq,
+            &info_seq,
+            DDS_LENGTH_UNLIMITED,
+            DDS_ANY_SAMPLE_STATE,
+            DDS_ANY_VIEW_STATE,
+            DDS_ANY_INSTANCE_STATE);
+    if (retcode == DDS_RETCODE_NO_DATA) {
+        return;
+    } else if (retcode != DDS_RETCODE_OK) {
+        fprintf(stderr, "take error %d\n", retcode);
+        return;
+    }
+
+    for (i = 0; i < DeviceStatusSeq_get_length(&data_seq); ++i) {
+        if (DDS_SampleInfoSeq_get_reference(&info_seq, i)->valid_data) {
```

Comment:
get the `sample_info*` as a variable, it makes it more readable (esp. after having to access it twice)